### PR TITLE
Null free lists

### PR DESCRIPTION
**Pull request by jakhog**
Introducing some mutable and immutable null free lists that we will use a few times in Runtime.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1115616103797671/1157679381642827)

### CODE REVIEW

File: Source/Collections/NullFreeList.cs [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        /// <inheritdoc/>
+        public void Clear()
+        {
+            _elements.Clear();
+        }
+
+        /// <inheritdoc/>
+        public bool Contains(T item)
+        {
+            if (item == null) return false;
```
Comment (woksin) [inline Source/Collections/NullFreeList.cs]: Do you have to handle null there, can’t you just rely on the call _elements.Contains(item)?

File: Source/Collections/NullFreeList.cs [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        /// <inheritdoc/>
+        public void CopyTo(T[] array, int arrayIndex)
+        {
+            _elements.CopyTo(array, arrayIndex);
+        }
+
+        /// <inheritdoc/>
+        public int IndexOf(T item)
+        {
+            if (item == null) return -1;
```
Comment (woksin) [inline Source/Collections/NullFreeList.cs]: Same as above

File: Source/Collections/NullFreeList.cs [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        public void Insert(int index, T item)
+        {
+            if (item == null) throw new ArgumentNullException(nameof(item));
+            _elements.Insert(index, item);
+        }
+
+        /// <inheritdoc/>
+        public bool Remove(T item)
+        {
+            if (item == null) return false;
```
Comment (woksin) [inline Source/Collections/NullFreeList.cs]: Same as above

File: Source/Collections/NullFreeList.cs (hunk 1 above)
Comment (woksin) [reply Source/Collections/NullFreeList.cs]: Doesn’t really matter and it is an optimization 😋

File: Source/Collections/NullFreeList.cs (hunk 3 above)
Comment (einari) [reply Source/Collections/NullFreeList.cs]: Would it make sense to throw  `ArgumentNullException` for all input parameters being Null on any method in these types of lists?   That would make it very consistent.

**Comment by einari** [comment]
Would it make sense to throw  `ArgumentNullException` for all input parameters being Null on any method in these types of lists?   That would make it very consistent.

File: Source/Collections/NullFreeList.cs (hunk 1 above)
Comment (jakhog) [reply Source/Collections/NullFreeList.cs]: It is a pure optimization yes - but so simple that I would like to keep it 🙂

**Comment by jakhog** [comment]
Not sure I quite understand. Do you mean the methods like `Contains`,`Remove`, etc. ?

**Comment by einari** [comment]
I was thinking about all methods taking an item as a parameter. Not sure it makes sense, but at least it would be very consistent - if you try to pass  `null`  to it in any way it would throw an exception.

**Comment by jakhog** [comment]
Hmm, it would make sense both ways - just different sense. The current implementation is permissive, so you could pass it along to other code reading it without worrying too much about exceptions. But that might not be necessary.

**Pull request merged**